### DISCUSSION TITLE
mfsmakesnapshot using "-c" and storage class problem

### DISCUSSION THREAD

**Discussion by BillKen** [moosefs/moosefs#420] [Q&A]
I have recently converted my mfs system to using storage class goals.  In the process I found ~580000 chunks in snapshots that still had a goal of 3 rather than the desired sclass of 3HDD in the web viewer.  For older snapshots its because the original files had a goal of 3 - that's fine but I was expecting snapshots made after moving the source to an sclass to inherrit that sclass but they were getting created with a goal of 3 instead.

It turns out that using "mfsmakesnapshot -c ..." creates snapshots that convert the sclass settings for the source files  to an equivalent goal where as when not using "-c" it works as expected.  The goal setting appears in the web view but doesnt showup when using utilities like mfsgetgoal or mfsgetsclass.

Is this a feature or a bug with the "-c" argument, or the viewer?
BillK

**Comment by chogata** [comment]
Storage class is assigned to files, not chunks. When a chunk starts to belong to many files - and the main reason for that to happen in MooseFS would be snapshots - there can come a time when it belongs to files with different storage classes. When that happens, the system has to decide something about how to store this chunk. Currently, if a chunk belongs to many different storage classes, a class with highest Redundancy Level (Goal equivalent - 1) is picked, but when there is more than one class with "maximum" RL, then the system just resorts to keeping the chunk in Goal RL+1, because there is no way to tell which storage class should be picked. So neither is.
So, in a way, both tools are right: mfsgetsclass will show you the storage class of the file(s).  CGI/CLI interface will show you, how actual chunks are stored...
We know this is not an ideal solution and we plan to change that in the future, probably by introducing Storage Classes priorities (then the class with highest priority will be picked for a chunk that belongs to different storage classes). But this is not yet 100% decided.
